templates/_worker.js/routes-matcher.ts

```diff
@@ -297,6 +298,87 @@ export class RoutesMatcher {
 		return prevPath;
 	}
 
+	/**
+	 * Applies the route's redirects for locales and internationalization.
+	 *
+	 * @param route Build output config source route.
+	 */
+	private applyLocaleRedirects(route: VercelSource): void {
+		if (!route.locale?.redirect) return;
+
+		if (!this.locales) this.locales = {};
+		Object.assign(this.locales, route.locale.redirect);
+
+		// Automatic locale detection is only supposed to occur at the root. However, the build output
+		// sometimes uses `/` as the regex instead of `^/$`. So, we should check if the `route.src` is
+		// equal to the path if it is not a regular expression, to determine if we are at the root.
+		// https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection
+		const srcIsRegex = /^\^(.)*$/.test(route.src);
+		if (!srcIsRegex && route.src !== this.path) return;
+
+		// If we already have a location header set, we might have found a locale redirect earlier.
+		if (this.headers.normal.has('location')) return;
+
+		const {
+			locale: { redirect: redirects, cookie: cookieName },
+		} = route;
+
+		const cookieValue = cookieName ? this.cookies[cookieName] ?? '' : '';
+		const cookieLocales = parseAcceptLanguage(cookieValue);
+
+		const headerLocales = parseAcceptLanguage(
+			this.reqCtx.request.headers.get('accept-language') ?? ''
+		);
+
+		// Locales from the cookie take precedence over the header.
+		const locales = [...cookieLocales, ...headerLocales];
+
+		const redirectLocales = locales
+			.map(locale => redirects[locale])
+			.filter(Boolean) as string[];
+
+		const redirectValue = redirectLocales[0];
+		if (redirectValue) {
+			const needsRedirecting = !this.path.startsWith(redirectValue);
+			if (needsRedirecting) {
+				this.headers.normal.set('location', redirectValue);
+				this.status = 307;
+			}
+			return;
+		}
+	}
+
+	/**
+	 * Modifies the source route's `src` regex to be friendly with previously found locale's in the
+	 * `miss` phase.
+	 *
+	 * Sometimes, there is a source route with `src: '/{locale}'`, which rewrites all paths containing
+	 * the locale to `/`. This is problematic for matching, and should only do this if the path is
+	 * exactly the locale, i.e. `^/{locale}$`.
```

Comment:
```suggestion
	 * Sometimes, there is a source route with `src: '/{locale}'`, which rewrites all paths containing
	 * the locale to `/`. This is problematic for matching, and should only do this if the path is
	 * exactly the locale, i.e. `^/{locale}$` (since as indicated in the Next.js docs:
	 * https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection
	 * and as we've confirmed via manual testing, only requests to the application's root should get redirected)
```

too verbose? :thinking:

Reply:
(PS: I might also be confusing this with the issue below? :thinking:)

Reply:
I don't actually think this is the right place to use that, as that pertains to the *hack* done for locale source routes from the `none` phase, meanwhile this is a *hack* to prevent it remapping to the wrong path with the stripped locale for matching non-locale assets/routes, if that makes sense.

Reply:
yeah makes sense, thanks :slightly_smiling_face::+1: